Issue 744 - Current Progress (checking about remove IO[Any,....])

TheMayoras opened the pull request:
I have not added Scaladoc links, moved traits to ZIO companion object, or organized in alphabetical order yet.

jdegoes commented:
@TheMayoras 

The pull request contains a _lot_ of files that shouldn't be there. I think maybe the IDE added them.

Anyway, before I review, could you remove the following files from the pull request:

 - All `.class` files / directories
 - All `.project` files / directories
 - All `.classpath` files / directories
 - All `.settings` files / directories

I think that'll clean it up quite a lot. 😅

TheMayoras commented:
@jdegoes My most recent commit has the files removed.

AshishPrasad commented:
@TheMayoras

The diff still contain `.prefs` and other non-scala files.

Probably, it is better to only include `.scala` files for this pull request.

LGLO commented:
@TheMayoras 
Take a look at diff here. It still contains a lot of changes.
I guess *exit* file should be removed and `sbt fmt` should be run.

TheMayoras commented:
@jdegoes @AshishPrasad I ran `sbt fmt` and removed a *lot* of extraneous files.  Hopefully it is a bit better.  Sorry for all the problems 😅

AshishPrasad commented:
@TheMayoras the `exit` file should not be added.

Also, `microsite` files should not be deleted.

---

I think, you can run `git show --name-only` and ensure only the intended files (typically `.scala` files) are added to the pull request.

jdegoes commented:
@TheMayoras Git is quite the beast to tame, don't let it stress you too much, and please give a holler if anything seems confusing.

jdegoes commented on core/jvm/src/test/scala/scalaz/zio/RTSSpec.scala:
```diff
@@ -578,15 +578,15 @@ class RTSSpec(implicit ee: ExecutionEnv) extends TestRuntime {
 
   def testAsyncIOEffectReturns =
-    unsafeRun(IO.effectAsyncM[Any, Throwable, Int](k => IO.effectTotal(k(IO.succeed(42))))) must_=== 42
+    unsafeRun(IO.effectAsyncM[Throwable, Int](k => IO.effectTotal(k(IO.succeed(42))))) must_=== 42
```
This are some really nice cleanups!

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,571 @@
+  // ALL error types in this trait must be a subtype of `UpperE`.
+  // TAKEN FROM ZIO_E_THROWABLE
+  // ALL environment types in this trait must be a supertype of `LowerR`.
+  type IoR = Any
```
I'd delete this, I don't think you need it anymore, or all the temporary comments above and below.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,571 @@
+
+  /** Accesses the whole environment of the effect.
+   */
+  final def environment: ZIO[IoR, Nothing, IoR] = ZIO.access(ZIO.identityFn[IoR])
```
Since we know `IoR` will be `Any`, we can delete `environment`, as it is not useless for the `IO` type.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,571 @@
+   * val portNumber = effect.access(_.config.portNumber)
+   * }}}
+   */
+  final def access: ZIO.AccessPartiallyApplied[IoR] =
```
Ditto for this, we know `R = Any`, so we can delete this method, since it won't be useful for `IO`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,571 @@
+
+  /** Effectfully accesses the environment of the effect.
+   */
+  final def accessM: ZIO.AccessMPartiallyApplied[IoR] =
```
Ditto for this, we know `R = Any`, so we can delete this method, since it won't be useful for `IO`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,571 @@
+   * This is similar to dependency injection, and the `provide` function can be
+   * thought of as `inject`.
+   */
+  final def provide[E, A](r: IoR): ZIO[IoR, E, A] => IO[E, A] =
```
Ditto for this, we know `R = Any`, so we can delete this method, since it won't be useful for `IO`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,571 @@
+   * (unsafely) execute tasks. This is useful for integration with
+   * non-functional code that must call back into functional code.
+   */
+  final def runtime: ZIO[IoR, Nothing, Runtime[IoR]] = ZIO.runtime
```
Instead of using `ZIO[IoR, ...`, try: `IO[...`, i.e. omit the `R` parameter entirely by using the `IO` type synonym.

In fact, that can help keep this file on track: all the effect types will be `IO` or weaker (e.g. `UIO` for finalizers, etc.).

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,571 @@
+   * This method can be used for terminating a fiber because a defect has been
+   * detected in the code.
+   */
+  final def die(t: Throwable): UIO[Nothing] = ZIO.die(t)
```
I'd delete all the comments in this file, later we can just link to ZIO documentation using the Scaladoc linking support:

```scala
/**
 * See [[scalaz.zio.ZIO.die]]
 */
```

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,571 @@
+   * stack. Manual use of this method can improve fairness, at the cost of
+   * overhead.
+   */
+  final val yieldNow: UIO[Unit] = ZIO.Yield
```
For all these files, try just to delegate to ZIO methods, not any of the data constructors directly.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,571 @@
+   * composite fiber that produces unit. This version is faster than [[forkAll]]
+   * in cases where the results of the forked fibers are not needed.
+   */
+  final def forkAll_[E, A](as: Iterable[ZIO[IoR, E, A]]): ZIO[IoR, Nothing, Unit] =
```
Remember to replace all the `ZIO[IoR, ...` in this file with `IO[...`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Promise.scala:
```diff
@@ -48,13 +46,12 @@ class Promise[E, A] private (private val state: AtomicReference[State[E, A]]) ex
-  final def await: IO[E, A] =
-    IO.effectAsyncInterrupt[Any, E, A](k => {
-      var result = null.asInstanceOf[Either[Canceler, IO[E, A]]]
+  final def await: IO[Any, E, A] =
```
This is a typo, `IO` will only support 2 type parameters `E` and `A`, not the `R` type parameter. Same as below.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Task.scala:
```diff
@@ -0,0 +1,631 @@
+  def apply[A](a: => A): Task[A] = effect(a)
+
+  type TaskE = Throwable
+  type TaskR = Any
```
I'd delete:

1. All the temp comments
2. The type aliases for `TaskE` and `TaskR`
3. Scaladoc for all the functions (instead, we can link to ZIO sScaladoc)

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Task.scala:
```diff
@@ -0,0 +1,631 @@
+   * The moral equivalent of `throw` for pure code.
+   * [[scalaz.zio.ZIO.fail]]
+   */
+  final def fail(error: TaskE): IO[TaskE, Nothing] = ZIO.fail(error)
```
Replace all `IO[???, A]` in this file by `Task[A]`, and delete all the environmental functions (`provide`, `accessM`, `access`, `environment`, etc.), because `R = Any`, they won't be useful.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Task.scala:
```diff
@@ -0,0 +1,631 @@
+   *
+   * [[scalaz.zio.ZIO.runtime]]
+   */
+  final def runtime: ZIO[TaskR, Nothing, Runtime[TaskR]] = ZIO.runtime
```
I'd replace all the `ZIO[TaskR, ...]` by either `Task` (if the error is `Throwable`) or `UIO` (if the error is `Nothing`).

jdegoes commented on core/shared/src/main/scala/scalaz/zio/TaskR.scala:
```diff
@@ -0,0 +1,566 @@
+ */
+object TaskR {
+
+  type TaskrE = Throwable
```
As in the other cases, I'd delete:

1. All the temp comments.
2. The `TaskrE` and `TaskR` type synonyms
3. All the Scaladoc (instead link to ZIO Scaladoc)

jdegoes commented on core/shared/src/main/scala/scalaz/zio/TaskR.scala:
```diff
@@ -0,0 +1,566 @@
+  /** Returns an effect that models failure with the specified error.
+   * The moral equivalent of `throw` for pure code.
+   */
+  final def fail(error: TaskrE): IO[TaskrE, Nothing] = ZIO.fail(error)
```
Instead of `IO[TaskrE, ...]`, use `Task[...]` everywhere in this file.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/TaskR.scala:
```diff
@@ -0,0 +1,566 @@
+
+  /** Accesses the whole environment of the effect.
+   */
+  final def environment[R]: ZIO[R, Nothing, R] = ZIO.access(ZIO.identityFn[R])
```
We don't really have a type alias for the combination but I see it a lot... we can fix it later.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/TaskR.scala:
```diff
@@ -0,0 +1,566 @@
+   * This is similar to dependency injection, and the `provide` function can be
+   * thought of as `inject`.
+   */
+  final def provide[R, A](r: R): ZIO[R, TaskrE, A] => IO[TaskrE, A] =
```
Replace `ZIO[R, TaskrE, ...]` by `TaskR[R, ...]` in this whole file.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/TaskR.scala:
```diff
@@ -0,0 +1,566 @@
+   * stack. Manual use of this method can improve fairness, at the cost of
+   * overhead.
+   */
+  final val yieldNow: UIO[Unit] = ZIO.Yield
```
Remember to delegate to the ZIO methods and not data constructors directly.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/TaskR.scala:
```diff
@@ -0,0 +1,566 @@
+  /** Returns an effect that will execute the specified effect fully on the
+   * provided executor, before returning to the default executor.
+   */
+  final def lock[R, A](executor: Executor)(zio: ZIO[R, TaskrE, A]): ZIO[R, TaskrE, A] =
```
This would be `TaskR[R, A]`, etc., here and elsewhere.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/TaskR.scala:
```diff
@@ -0,0 +1,566 @@
+   * the more expressive variant of this function that can return a value
+   * synchronously.
+   */
+  final def effectAsync[A](register: (ZIO[Any, TaskrE, A] => Unit) => Unit): ZIO[Any, TaskrE, A] =
```
In the rare situation where `ZIO[Any, TaskrE, ...]` appears, that would translate to `Task[A]`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/UIO.scala:
```diff
@@ -0,0 +1,534 @@
+object UIO {
+
+  type UioE = Nothing
+  type UioR = Any
```
As in the other cases, I'd delete:

1. Temp comments
2. Scaladoc (link to ZIO Scaladoc)
3. Type aliases `UioE` and `UioR`
4. Because `UIO` doesn't use environment, delete the environmental functions (`provide`, `access`, `accessM`, `environment`)

jdegoes commented on core/shared/src/main/scala/scalaz/zio/UIO.scala:
```diff
@@ -0,0 +1,534 @@
+  /** Returns an effect that models failure with the specified error.
+   * The moral equivalent of `throw` for pure code.
+   */
+  final def fail(error: UioE): IO[UioE, Nothing] = ZIO.fail(error)
```
Replace all the `IO[UioE, ...]` with `UIO[...]`.

jdegoes replied:
Ah, we can also delete `fail` since the user cannot call the method (no values of type `Nothing` exist). 😄

jdegoes commented on core/shared/src/main/scala/scalaz/zio/UIO.scala:
```diff
@@ -0,0 +1,534 @@
+   * supervised effect completes.
+   */
+  final def superviseWith[A](
+    zio: ZIO[UioR, UioE, A]
```
Ditto for all these, they become much simpler, e.g. `UIO[A]` in the above example.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/UIO.scala:
```diff
@@ -0,0 +1,534 @@
+
+  /** Returns an effectful function that merely swaps the elements in a `Tuple2`.
+   *
+   *  Q: Leave Implicit?
```
Actually we should delete any methods like `swap` which assume or require that `R` be anything other than `Any`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/UIO.scala:
```diff
@@ -0,0 +1,534 @@
+  /** Returns an effectful function that extracts out the second element of a
+   * tuple.
+   */
+  final def _2[A, B](implicit ev: UioR <:< (A, B)): ZIO[UioR, UioE, B] = ZIO._2
```
Yes, both these two must be deleted, not just in `UIO.scala`, but also in `Task.scala`, because they assume / require that `R` be some type other than `Any`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/UIO.scala:
```diff
@@ -0,0 +1,534 @@
+  /** Requires that the given `IO[E, Option[A]]` contain a value. If there is no
+   * value, then the specified error will be raised.
+   */
+  final def require[A](error: UioE): IO[UioE, Option[A]] => IO[UioE, A] =
```
This can be deleted because `UIO` cannot fail.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/UIO.scala:
```diff
@@ -0,0 +1,534 @@
+   */
+  final def bracketExit[A, B](
+    acquire: ZIO[UioR, UioE, A],
+    release: (A, Exit[UioE, B]) => ZIO[UioR, Nothing, _],
```
You can use `Exit[Nothing, B]`, etc. so we can delete `UioE`.

jdegoes commented on exit:
```diff
@@ -0,0 +1,636 @@
+[33mcommit ce58772a009a4cef29f1eab4581bc6ccc9e7d964[m[33m ([m[1;36mHEAD -> [m[1;32mmaster[m[33m, [m[1;31morigin/master[m[33m, [m[1;31morigin/HEAD[m[33m)[m
+Author: TheMayoras <ben.mayoras@gmail.com>
```
This file should not be added to the commit.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZIO.scala:
```diff
@@ -1078,22 +1078,27 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
+ * Traits that IO, UIO, TaskR, and Task Inherited.
+ * ------------------------------------------------------------------------------------------------
+ * ------------------------------------------------------------------------------------------------
+ * ------------------------------------------------------------------------------------------------
```
I'd delete all these temp comments.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZIO.scala:
```diff
@@ -1078,22 +1078,27 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
    * The moral equivalent of `throw` for pure code.
    */
-  final def fail[E <: UpperE](error: E): IO[E, Nothing] = halt(Cause.fail(error))
+  final def fail[E](error: E): IO[E, Nothing] = halt(Cause.fail(error))
```
HUGE and beautiful cleanups here and elsewhere!!! 😍

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZIO.scala:
```diff
@@ -1826,29 +1828,47 @@ private[zio] trait ZIO_R_Any extends ZIO_E_Any {
-
-  def apply[A](a: => A): UIO[A] = effectTotal(a)
-}
+/*
```
I'd delete all these temp comments.

jdegoes commented on microsite/src/main/resources/microsite/css/override.css:
```diff
@@ -1,75 +0,0 @@
-.jumbotron {
```
This was accidentally deleted and should not be deleted.

jdegoes commented:
@TheMayoras This is coming along great. Please let me know if you have any questions. I'd love to get this in for RC5, right before we switch over to the new Github organization (zio).

TheMayoras commented:
@jdegoes I haven't had much time to work on it this week (finals just finished up), but I should start vamping up how much I can work on it.  When does RC5 start?

jdegoes commented:
@TheMayoras No worries—thanks for squeezing OSS contributions in between classes / tests. 😄 

RC5 will come out early this week (Monday / Tuesday).

TheMayoras commented:
@jdegoes Just wanted to give you an update.  I think all I have left is sorting all of the methods.  I am organizing the methods in ZIO, IO, UIO, TaskR, and Task alphabetically correct?

jdegoes commented:
@TheMayoras Yes, that's right!

TheMayoras commented:
@jdegoes I believe I have almost everything ready to go!

TheMayoras commented:
@jdegoes I just wanted to make sure you saw my most recent pull request!

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,357 @@
+
+  /** See [[scalaz.zio.ZIO.unit]]
+   */
+  final val unit: IO[Nothing, Unit] = ZIO.unit
```
`UIO[Unit]`

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,357 @@
+import scalaz.zio.internal.{ Executor, Platform }
+import scala.concurrent.ExecutionContext
+
+object IO {
```
Can you please place all methods in alphabetical ordering for easy lookup / organization? Thank you!

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,357 @@
+
+  /** See [[scalaz.zio.ZIO.bracket]]
+   */
+  final def bracket[E, A, B](acquire: IO[E, A], release: A => IO[Nothing, _], use: A => IO[E, B]): IO[E, B] =
```
For release: `UIO[_]`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,357 @@
+   */
+  final def bracketExit[E, A, B](
+    acquire: IO[E, A],
+    release: (A, Exit[E, B]) => IO[Nothing, _],
```
For release: `(A, Exit[E, B]) => UIO[_]`

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,357 @@
+  final def collectAllParN[E, A](n: Long)(as: Iterable[IO[E, A]]): IO[E, List[A]] =
+    ZIO.collectAllParN(n)(as)
+
+  /** See [[scalaz.zio.ZIO.ZIO.die]]
```
I think this should be: `[[scalaz.zio.ZIO.die]]`

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,357 @@
+
+  /** See [[scalaz.zio.ZIO.effectAsyncM]]
+   */
+  final def effectAsyncM[E, A](register: (IO[E, A] => Unit) => IO[Nothing, _]): IO[E, A] =
```
`register: (IO[E, A] => Unit) => UIO[_]`

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,357 @@
+
+  /** See [[scalaz.zio.ZIO.fromFunction]]
+   */
+  final def fromFunction[A](f: Any => A): IO[Nothing, A] =
```
`: UIO[A] =`

TheMayoras replied:
Should all IO[Nothing, _] be changed to UIO[_]?

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,357 @@
+
+  /** See [[scalaz.zio.ZIO.identity]]
+   */
+  final def identity: IO[Nothing, Any] = ZIO.identity
```
This function should be deleted, it only makes sense for effect types with `R`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,357 @@
+
+  /** See [[scalaz.zio.ZIO.lock]]
+   */
+  final def lock[R, E, A](executor: Executor)(zio: IO[E, A]): ZIO[R, E, A] =
```
There will be no `R` type parameter here, and `lock` will return `IO[E, A]`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,357 @@
+
+  /** See [[scalaz.zio.ZIO.runtime]]
+   */
+  final def runtime: IO[Nothing, Runtime[Any]] = ZIO.runtime
```
`: UIO[Runtime[Any]]`

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,357 @@
+
+  /** See [[scalaz.zio.ZIO.supervisedWith]]
+   */
+  final def superviseWith[E, A](zio: IO[E, A])(supervisor: IndexedSeq[Fiber[_, _]] => IO[Nothing, _]): IO[E, A] =
```
`supervisor: IndexedSeq[Fiber[_, _]] => UIO[_]`

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,357 @@
+
+  /** See [[scalaz.zio.ZIO.supervised]]
+   */
+  def supervised[E, A](zio: IO[E, A]): IO[E, A] =
```
It would be good to "search and replace" all `zio` parameter names with `io` parameter names. 👍

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,357 @@
+
+  /** See [[scalaz.zio.ZIO.swap]]
+   */
+  final def swap[E, A, B](implicit ev: Any <:< (A, B)): IO[E, (B, A)] =
```
This function only makes sense when `R` != `Any`, so should be deleted from `IO` object.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,357 @@
+
+  /** See [[scalaz.zio.ZIO._1]]
+   */
+  final def _1[E, A, B](implicit ev: Any <:< (A, B)): IO[E, A] = ZIO._1
```
This function only makes sense when `R` != `Any` and should be deleted from `IO`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,357 @@
+
+  /** See [[scalaz.zio.ZIO._2]]
+   */
+  final def _2[E, A, B](implicit ev: Any <:< (A, B)): IO[E, B] = ZIO._2
```
This function only makes sense when `R` != `Any` and should be deleted from `IO`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Promise.scala:
```diff
@@ -20,40 +20,37 @@ import java.util.concurrent.atomic.AtomicReference
- * A promise represents an asynchronous variable that can be set exactly once,
- * with the ability for an arbitrary number of fibers to suspend (by calling
- * `get`) and automatically resume when the variable is set.
+/** A promise represents an asynchronous variable that can be set exactly once,
```
I'd add a newline right after `/**`, because well-formatted Scaladoc will start on the next line, after the initial `/**`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Promise.scala:
```diff
@@ -20,40 +20,37 @@ import java.util.concurrent.atomic.AtomicReference
-  /**
-   * Checks for completion of this Promise. Produces true if this promise has
-   * already been completed with a value or an error and false otherwise.
+  /** Checks for completion of this Promise. Produces true if this promise has
```
Same here.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Promise.scala:
```diff
@@ -20,40 +20,37 @@ import java.util.concurrent.atomic.AtomicReference
- * higher-level concurrent or asynchronous structures.
- * {{{
- * for {
+ *  Promises can be used for building primitive actions whose completions
```
Extra initial spaces, just 1 space is required between `*` and the first character of the comment.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Promise.scala:
```diff
@@ -20,40 +20,37 @@ import java.util.concurrent.atomic.AtomicReference
-  /**
-   * Retrieves the value of the promise, suspending the fiber running the action
-   * until the result is available.
+  /** Retrieves the value of the promise, suspending the fiber running the action
```
Same here.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Promise.scala:
```diff
@@ -77,35 +74,30 @@ class Promise[E, A] private (private val state: AtomicReference[State[E, A]]) ex
 
-  /**
-   * Completes immediately this promise and returns optionally it's result.
+  /** Completes immediately this promise and returns optionally it's result.
```
Ditto here.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Promise.scala:
```diff
@@ -77,35 +74,30 @@ class Promise[E, A] private (private val state: AtomicReference[State[E, A]]) ex
 
-  /**
-   * Completes the promise with the specified value.
+  /** Completes the promise with the specified value.
```
And here.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Promise.scala:
```diff
@@ -77,35 +74,30 @@ class Promise[E, A] private (private val state: AtomicReference[State[E, A]]) ex
-  /**
-   * Fails the promise with the specified error, which will be propagated to all
-   * fibers waiting on the value of the promise.
+  /** Fails the promise with the specified error, which will be propagated to all
```
And here.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Promise.scala:
```diff
@@ -77,35 +74,30 @@ class Promise[E, A] private (private val state: AtomicReference[State[E, A]]) ex
-  /**
-   * Completes the promise with interruption. This will interrupt all fibers
-   * waiting on the value of the promise.
+  /** Completes the promise with interruption. This will interrupt all fibers
```
And here...

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Promise.scala:
```diff
@@ -77,35 +74,30 @@ class Promise[E, A] private (private val state: AtomicReference[State[E, A]]) ex
-  /**
-   * Completes the promise with the specified result. If the specified promise
-   * has already been completed, the method will produce false.
+  /** Completes the promise with the specified result. If the specified promise
```
And here...

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Promise.scala:
```diff
@@ -176,24 +168,20 @@ class Promise[E, A] private (private val state: AtomicReference[State[E, A]]) ex
 
-  /**
-   * Makes a new promise.
+  /** Makes a new promise.
```
I would change your IntelliJ IDEA configuration to insert a newline after `/**`, then you can just have it format the Scaladoc; you can also make IntelliJ IDEA use the `.scalafmt` so it knows the "right" way to format code.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Promise.scala:
```diff
@@ -176,24 +168,20 @@ class Promise[E, A] private (private val state: AtomicReference[State[E, A]]) ex
-   * Acquires a resource and performs a state change atomically, and then
-   * guarantees that if the resource is acquired (and the state changed), a
-   * release action will be called.
+  /** Acquires a resource and performs a state change atomically, and then
```
More Scaladoc.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Task.scala:
```diff
@@ -0,0 +1,341 @@
+
+object Task {
+
+  /** See [[scalaz.zio.ZIO.interrupt]]
```
All the Scaladoc in all the files looks like it's missing the initial newline after `/**`, I think a quick change to config should fix it! 👍

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Task.scala:
```diff
@@ -0,0 +1,341 @@
+
+import scala.concurrent.ExecutionContext
+
+object Task {
```
I would alphabetize all the methods in this calss for easy lookup / better organization.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Task.scala:
```diff
@@ -0,0 +1,341 @@
+
+  /** See [[scalaz.zio.ZIO.fail]]
+   */
+  final def fail(error: Throwable): IO[Throwable, Nothing] = ZIO.fail(error)
```
`: Task[Nothing]`

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Task.scala:
```diff
@@ -0,0 +1,341 @@
+
+  /** See [[scalaz.zio.ZIO.fromFunction]]
+   */
+  final def fromFunction[A](f: Any => A): UIO[A] =
```
This one only makes sense when `R` != `Any`, and it should be deleted from `Task`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Task.scala:
```diff
@@ -0,0 +1,341 @@
+
+  /** See [[scalaz.zio.ZIO.fromFunctionM]]
+   */
+  final def fromFunctionM[A](f: Any => IO[Throwable, A]): Task[A] =
```
This one only makes sense when `R` != `Any`, and it should be deleted from `Task`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -0,0 +1,357 @@
+
+  /** See [[scalaz.zio.ZIO.fromFunctionM]]
+   */
+  final def fromFunctionM[E, A](f: Any => IO[E, A]): IO[E, A] =
```
This one only makes sense when `R` != `Any`, and it should be deleted from `IO`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Task.scala:
```diff
@@ -0,0 +1,341 @@
+
+  /** See [[scalaz.zio.ZIO.fromEither]]
+   */
+  final def fromEither[A](v: => Either[Throwable, A]): IO[Throwable, A] =
```
`: Task[A]`

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Task.scala:
```diff
@@ -0,0 +1,341 @@
+
+  /** See [[scalaz.zio.ZIO.fromFiber]]
+   */
+  final def fromFiber[A](fiber: => Fiber[Throwable, A]): IO[Throwable, A] =
```
`: Task[A]`

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Task.scala:
```diff
@@ -0,0 +1,341 @@
+
+  /** See [[scalaz.zio.ZIO.fromFiberM]]
+   */
+  final def fromFiberM[A](fiber: IO[Throwable, Fiber[Throwable, A]]): IO[Throwable, A] =
```
`def fromFiberM[A](fiber: Task[Fiber[Throwable, A]]): Task[A]`

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Task.scala:
```diff
@@ -0,0 +1,341 @@
+
+  /** See [[scalaz.zio.ZIO.halt]]
+   */
+  final def halt(cause: Cause[Throwable]): IO[Throwable, Nothing] = ZIO.halt(cause)
```
`: Task[Nothing]`

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Task.scala:
```diff
@@ -0,0 +1,341 @@
+
+  /** See [[scalaz.zio.ZIO.identity]]
+   */
+  final def identity: UIO[Any] = ZIO.identity
```
This method only makes sense when `R` != `Any`, and should be deleted from `Task`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Task.scala:
```diff
@@ -0,0 +1,341 @@
+
+  /** See [[scalaz.zio.ZIO.require]]
+   */
+  final def require[A](error: Throwable): IO[Throwable, Option[A]] => IO[Throwable, A] =
```
`: Task[Option[A]]`

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Task.scala:
```diff
@@ -0,0 +1,341 @@
+
+  /** See [[scalaz.zio.ZIO.unsandbox]]
+   */
+  final def unsandbox[A](v: IO[Cause[Throwable], A]): Task[A] = ZIO.unsandbox(v)
```
All the methods of the `Task` object should accept and return types compatible with `Task`. The type `IO[Cause[Throwable], A]` is not compatible with `Task`, because `Cause[Throwable]` is not a subtype of `Throwable`. So we should remove this method from the `Task` object.

jdegoes replied:
It looks like the old code got this wrong too!

jdegoes commented on core/shared/src/main/scala/scalaz/zio/TaskR.scala:
```diff
@@ -0,0 +1,370 @@
+
+import scala.concurrent.ExecutionContext
+
+object TaskR {
```
I would alphabetize all the methods in this object for better organization.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Task.scala:
```diff
@@ -0,0 +1,341 @@
+
+  /** See [[scalaz.zio.ZIO.when]]
+   */
+  final def when(b: Boolean)(zio: Task[_]): Task[Unit] =
```
I would do a search / replace and rename all `zio` parameters to `task` parameters.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/TaskR.scala:
```diff
@@ -0,0 +1,370 @@
+
+  /** See [[scalaz.zio.ZIO.effectAsyncMaybe]]
+   */
+  final def effectAsyncMaybe[A](register: (Task[A] => Unit) => Option[Task[A]]): Task[A] =
```
I think this one can have the `R` parameter and use `TaskR[R, A]` (a change by @iravid made this possible).

jdegoes commented on core/shared/src/main/scala/scalaz/zio/TaskR.scala:
```diff
@@ -0,0 +1,370 @@
+
+  /** See [[scalaz.zio.ZIO.require]]
+   */
+  final def require[A](error: Throwable): Task[Option[A]] => Task[A] =
```
Should be polymorphic in `R`:

```scala
def require[R, A](error: Throwable): TaskR[R, Option[A]] => TaskR[R, A]`
```

jdegoes commented on core/shared/src/main/scala/scalaz/zio/TaskR.scala:
```diff
@@ -0,0 +1,370 @@
+
+  /** See [[scalaz.zio.ZIO.unsandbox]]
+   */
+  final def unsandbox[R, A](v: ZIO[R, Cause[Throwable], A]): TaskR[R, A] = ZIO.unsandbox(v)
```
I would delete this method from `TaskR`, because the input value `v` cannot be typed as a `TaskR`, since `Cause[Throwable]` is not a subtype of `Throwable`. If a user wants to use this function, they have to go to `IO` or `ZIO`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/TaskR.scala:
```diff
@@ -0,0 +1,370 @@
+
+  /** See [[scalaz.zio.ZIO._2]]
+   */
+  final def _2[R, A, B](implicit ev: R <:< (A, B)): TaskR[R, B] = ZIO._2
```
I'd search and replace all parameter names `zio` and `io` to `taskr`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/UIO.scala:
```diff
@@ -0,0 +1,285 @@
+import scalaz.zio.Exit.Cause
+import scalaz.zio.internal.{ Executor, Platform }
+
+object UIO {
```
I'd alphabetize all method names in this object.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/UIO.scala:
```diff
@@ -0,0 +1,285 @@
+
+  /** See [[scalaz.zio.ZIO.identity]]
+   */
+  final def identity: UIO[Any] = ZIO.identity
```
This function only makes sense when `R` != `Any`, so should be deleted from `UIO`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/UIO.scala:
```diff
@@ -0,0 +1,285 @@
+
+  /** See [[scalaz.zio.ZIO.raceAll]]
+   */
+  final def raceAll[R1 <: Any, A](zio: UIO[A], ios: Iterable[ZIO[R1, Nothing, A]]): ZIO[R1, Nothing, A] =
```
Delete `R1` and change `ios` to `Iterable[UIO[A]]`, with return value `UIO[A]`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/UIO.scala:
```diff
@@ -0,0 +1,285 @@
+
+  /** See [[scalaz.zio.ZIO.reduceAll]]
+   */
+  final def reduceAll[R1 <: Any, A](a: UIO[A], as: Iterable[ZIO[R1, Nothing, A]])(f: (A, A) => A): ZIO[R1, Nothing, A] =
```
Delete `R1`, change `as` to `Iterable[UIO[A]]`, and return `UIO[A]`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/UIO.scala:
```diff
@@ -0,0 +1,285 @@
+
+  /** See [[scalaz.zio.ZIO.reduceAllPar]]
+   */
+  final def reduceAllPar[R1 <: Any, A](a: UIO[A], as: Iterable[ZIO[R1, Nothing, A]])(
```
Delete `R1`, change `as` to `Iterable[UIO[A]]`, and return `UIO[A]`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/UIO.scala:
```diff
@@ -0,0 +1,285 @@
+
+  /** See [[scalaz.zio.ZIO.unsandbox]]
+   */
+  final def unsandbox[A](v: IO[Cause[Nothing], A]): UIO[A] = ZIO.unsandbox(v)
```
I'd delete `unsandbox` from `UIO` because the input to this function cannot be typed as `UIO`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/UIO.scala:
```diff
@@ -0,0 +1,285 @@
+
+  /** See [[scalaz.zio.ZIO.whenM]]
+   */
+  final def whenM(b: UIO[Boolean])(zio: UIO[_]): UIO[Unit] =
```
I'd search and replace all `io` / `zio` parameter names to `uio` name.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZIO.scala:
```diff
@@ -1078,219 +1078,219 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
-  type UpperE
-  // ALL environment types in this trait must be a supertype of `LowerR`.
-  type LowerR
+object ZIO {
```
`object ZIO extends Serializable`

jdegoes replied:
This is a huge cleanup, by the way! It's going to make the Scaladoc look much better and improve type inference / error messages everywhere.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZIO.scala:
```diff
@@ -1078,219 +1078,219 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
+   * Returns an effect that is interrupted.
    */
-  final def halt[E <: UpperE](cause: Cause[E]): IO[E, Nothing] = new ZIO.Fail(cause)
+  final val interrupt: UIO[Nothing] = halt(Cause.interrupt)
```
I'd alphabetize all methods inside the `ZIO` object for better organization.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZIO.scala:
```diff
@@ -1078,219 +1078,219 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
    * synchronously.
    */
-  final def effectAsync[E <: UpperE, A](register: (ZIO[Any, E, A] => Unit) => Unit): ZIO[Any, E, A] =
+  final def effectAsync[E, A](register: (ZIO[Any, E, A] => Unit) => Unit): ZIO[Any, E, A] =
```
Thanks to @iravid's change, I think this can now use `R`. Be sure to copy over his changes and any other new ones when you pull in latest changes from master.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZIO.scala:
```diff
@@ -1301,14 +1301,14 @@ private[zio] trait ZIOFunctions extends Serializable {
    * the value synchronously.
    */
-  final def effectAsyncMaybe[E <: UpperE, A](register: (ZIO[Any, E, A] => Unit) => Option[IO[E, A]]): ZIO[Any, E, A] =
+  final def effectAsyncMaybe[E, A](register: (ZIO[Any, E, A] => Unit) => Option[IO[E, A]]): ZIO[Any, E, A] =
```
This can now be polymorphic in `R` too, I think.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZIO.scala:
```diff
@@ -1632,226 +1662,154 @@ private[zio] trait ZIOFunctions extends Serializable {
-    in: Iterable[ZIO[R, E, A]]
-  )(zero: B)(f: (B, A) => B): ZIO[R, E, B] =
-    in.foldLeft[ZIO[R, E, B]](succeedLazy[B](zero))((acc, a) => acc.zipPar(a).map(f.tupled))
+  final def require[E, A](error: E): IO[E, Option[A]] => IO[E, A] =
```
This can be generalized to:

```scala
final def require[R, E, A](error: E): ZIO[R, E, Option[A]] => ZIO[R, E, A]
```

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZIO.scala:
```diff
@@ -1632,226 +1662,154 @@ private[zio] trait ZIOFunctions extends Serializable {
    */
-  final def whenM[R >: LowerR, E <: UpperE](b: ZIO[R, E, Boolean])(zio: ZIO[R, E, _]): ZIO[R, E, Unit] =
-    b.flatMap(b => if (b) zio.unit else unit)
+  final def sleep(duration: Duration): ZIO[Clock, Nothing, Unit] =
```
I would add the `sleep` method to `TaskR`, as well, because it is compatible with `TaskR` type.

jdegoes commented:
@TheMayoras 

Once you've made these changes, there are a couple additional steps to take to make the pull request mergeable:

1. Pull in the latest changes from master. This will result in _merge conflicts_, because the `ZIO.scala` file has been modified since you started. The modifications are likely around @iravid's changes to add `R` to various methods, and you should be careful to copy them in.
2. Run `fmt` from SBT to format all the source code files.
3. Finally, run `makeMicrosite` to make the microsite. You might see some errors in the Markdown files. Most likely caused by removing a type parameter `R` from different methods (e.g. `IO.effectAsyncInterrupt`). These should be quick fixes.

Please let me know if you need any help!

TheMayoras commented:
@jdegoes I am probably totally missing it, but I'm not sure where the "makeMicrosite" .exe is

jdegoes commented:
@TheMayoras It's an `sbt` command. `sbt makeMicrosite`. See [here](https://github.com/scalaz/scalaz-zio/blob/master/CONTRIBUTING.md) for more details.


> @jdegoes Should anonymous function be changed also? Changing to ZIO[R, E, Option[A]] => ZIO[R, E, A] gives me a type error because (io: IO[E, Option[A]]) causes a type error, but zio: ZIO[R, E, Option[A]] resolves the type error.

Ah yes, that's good! 👍 

Anything else I can help with?

TheMayoras commented:
@jdegoes Would you like the alphabetical sorting over everything or withing groups?  For example, all the `val` grouped and sorted.

jdegoes commented:
@TheMayoras I would do it over everything, e.g. `def` and `val` all mixed together, rather than separated.

jdegoes commented:
@TheMayoras Looks like there are more updates that have to be merged in manually. Let me know if you need any help on that!

TheMayoras commented:
@jdegoes I'm having some issues I'm hoping you can help with.  I am missing a class in the ZIO object called `EffectTotalWith[A]`.  It seems to me one of the newer merges removed that class.  Should I remove effectTotalWith from all of the new files (IO, UIO, Task, and TaskR)?

jdegoes commented:
> @jdegoes I'm having some issues I'm hoping you can help with. I am missing a class in the ZIO object called EffectTotalWith[A]. It seems to me one of the newer merges removed that class. Should I remove effectTotalWith from all of the new files (IO, UIO, Task, and TaskR)?

Yes, please do. This was replaced by `suspendWith`, which should be added to all the new files.

TheMayoras commented:
@jdegoes I'm also running into issues with 
`  final def runtime[R]: ZIO[R, Nothing, Runtime[R]] =
    for {
      environment <- environment[R]
      platform    <- effectTotalWith(ZIO.identityFn[Platform])
    } yield Runtime(environment, platform)`.

If I swap out `platform <-effectTotalWith` with `platform <-suspendWith` I get a type mismatch.

jdegoes commented:
See the new implementation of `runtime` in ZIO.scala.

TheMayoras commented:
@jdegoes I think it is very close.  The only problem I have right now is that `FiberContext.scala` uses a `UIO.foreach_`, but I don't have `foreach_` in `UIO`.  I could add it, but I just want to make sure it's correct.  I ran `git merge` this morning, but I still have to run `sbt fmt` and `sbt makeMicrosite`.

jdegoes commented:
> The only problem I have right now is that FiberContext.scala uses a UIO.foreach_, but I don't have foreach_ in UIO. I could add it, but I just want to make sure it's correct. 

Yes, please add it. Keep in mind the version for `UIO` won't support any errors in the effects, or any `R`, either, and will just delegate to `ZIO.foreach_`.

> I ran git merge this morning, but I still have to run sbt fmt and sbt makeMicrosite.

We switched over to Docusaurus so the commands changed a bit (see the CONTRIBUTING.md file).

> I think it is very close.

Yes indeed, I'm hoping for tomorrow! 🎉

TheMayoras commented:
@jdegoes I'm now having issues with 
`scalaz-zio\core\jvm\src\main\java\scalaz\zio\internal\impls\padding\MutableQueueFieldsPadding`
It's failing to compile stating that it is out of heap space and cannot find `MutableConcurrentQueue`.  It's imported at the top, but the actual file is in the shared directory not the JVM directory.  

I have tried upping SBT's heap space, but it does not seem to help.  Am I totally missing something?

jdegoes commented:
@TheMayoras It looks like in a recent merge, some new changes from ZIO.scala were not successfully integrated. The build is now failing because the different source files are not compilable (they assume the existence of code which does not exist in this version).

At this point, the safest way to ensure you have all changes may be to: take your UIO.scala, TaskR.scala, Task.scala, and IO.scala files, and add them to a _fresh checkout_, and then redo the work on ZIO.scala, starting from the latest ZIO.scala. If you need some help, I am happy to help!

jdegoes commented:
> I have tried upping SBT's heap space, but it does not seem to help. Am I totally missing something?

Try running using the sbt script inside the project (`sbt`), which is configured with enough heap space to reliably build the project.

TheMayoras commented:
@jdegoes So I have copied over the current `ZIO.scala` file and updated it to where my previous version was at.  I've tried running `./sbt`, `\sbt`, and `.\sbt` on the terminal, but I cannot get it to work.  If I try running `./sbt` on the Git Bash, I get an error stating I have an `Invalid maximum heap size` not matter how small I try to make it.  If I compile with the normal `sbt` command, I get the same error about not finding `MutableConcurrentQueue`.

TheMayoras commented:
I am on Windows btw

jdegoes commented:
@TheMayoras How much memory does your machine have? I found [these instructions](https://javarevisited.blogspot.com/2012/12/invalid-initial-and-maximum-heap-size.html) for this error message.

TheMayoras commented:
@jdegoes I have 8G RAM.

TheMayoras commented:
@jdegoes It's now throwing 
`[error] Error while emitting Chunk.scala`
`[error] Java heap space`
when compiling...?

TheMayoras commented:
@jdegoes So I'm not quite sure what my issue is with compiling, but I am going to start from a new fork and retry if that works with you.

jdegoes commented:
@TheMayoras Sorry, I have been incredibly busy and will remain so until after LambdaConf. 

It's probably ok to start with a clean fork (copying over your new files, which should be safe, aside from the package name changes `scalaz.zio` => `zio`). Also I'd try increasing maximum heap to maybe 7 GB, since you have an 8 GB machine. Finally I'd close as many applications as possible while you're compiling.

I would _love_ for this PR to be included in RC6 which should come out next weekend if we get the build situation lined up, and happy to help commit to your fork if necessary to cross the finish line. 

This one's going to make a big difference in usability and type inference.

jdegoes commented:
@TheMayoras 

I'd like [here](https://stackoverflow.com/questions/17369522/set-default-heap-size-in-windows), as well as maybe locally edit `sbt` script (I assume you're running some sort of Git bash or something to get that working) and increase max heap that way.

There is also `.sbtopts` that could be tweaked. Some combination of editing all of these may help.

I'd also edit with Metals / VS Code and not IntelliJ IDEA, which will consume many GB by itself.

TheMayoras commented:
@jdegoes IT FINALLY COMPILES.  You were totally right, I had to adjust my .sbtops and the SBT_OPTS & JAVA_OPTS that I thought I was using were not actually getting used 😅.  So now I am installing Metals for VS Code.  Now I'm just working on some issues with `RTSS` and `ZManaged` throwing error with `A type was inferred to be any.`

TheMayoras commented:
@jdegoes I am having problems with Metals finding the `zio.internal` package, specifically the `stacktracer` package.  
The only other problem is with some methods in `ZIO`.  For example, on 
`checkSupervised(SuperviseStatus =>ZIO[R, E, A])` 
I get `class SuperviseStatus takes type parameters`.  If I then change to `SuperviseStatus[R, E, A]`, I then get: 
![image](https://user-images.githubusercontent.com/46846262/58921505-16494e80-8705-11e9-91ac-5dede562fee0.png)

`SuperviseStatus` is currently a member of the `ZIO` object.

TheMayoras commented:
@jdegoes 
So for a status update:
I created a fresh clone of current build, and added my companion files to them.  It seems I'm missing some newer functions, which is a bummer...I have added the functions to the companion objects to get it to compile, but I need to go through more carefully to ensure all of the functions are in the companion objects. 

`compile` runs without errors, but I am still getting the `type was inferred to be Any` error.

I need to re-alphabetize the new `ZIO.scala` file.

My current commit is the repo `zio` on my profile.  

My main concern at this point is that I'm missing some important functions in the companion files.

TheMayoras commented:
@jdegoes 
Could you help me out a big when you get a chance?  I'm really at a loss. 😕

Thank you

jdegoes commented:
@TheMayoras 

What time zone you in? Any chance I can help tomorrow?

TheMayoras commented:
@jdegoes Yes!  I'm eastern time zone!  I'm ready to finally finish this.  I feel bad its taken me so long 😓.  What times work best for you?

jdegoes commented:
9:30am?

TheMayoras commented:
@jdegoes Works for me.  I have something at noon tomorrow, so I can always meet earlier if need be.  How would you like to do this?

TheMayoras commented:
@jdegoes how would you like to meet?

jdegoes commented:
@TheMayoras 

https://zoom.us/j/117294361

TheMayoras commented:
@jdegoes 

All alphabetized!

I've run `./sbt test` and `./sbt docs/docusaurusCreateSite`.  
I'm running out of threads for `test` and `docs...` is telling me I don't have `yarn`.  

All the methods are formatted and it is compiling though.

jdegoes commented:
@TheMayoras There are a few conflicts in ZIO.scala. That probably just means there was a few changes since the last pull. Can you inspect and copy over the changes?

The build failed, but as far as I can see, but it's probably because the code needs to be formatted.

From `sbt`, can you run `fmt`? This will format all the source code and get it ready to merge!

TheMayoras commented:
@jdegoes 
Just `git pull`'d and everything compiles.  I ran `./sbt fmt`, but `docs/docusaurusCreateSite` still errors out about not being able to run `yarn`.

jdegoes commented:
You need to install `nodejs` and `yarn` globally on Windows. If you have a package manager you can do that, otherwise you will need installers. 

However, that should not be affecting the failed build. Did you push changes? Let me take a look...

TheMayoras commented:
@jdegoes 
Installing yarn and nodejs right now.  

Apparently I pushed to the wrong repo 😅.  This should be the latest push now.

TheMayoras commented:
@jdegoes 
So I have yarn and nodejs, but I still can't run `docs/docusarusCreateSite`; it gives me the same error.  I can run `yarn` straight from the git bash and cmd.exe, so I'm not sure what to make of this.

jdegoes commented:
@TheMayoras It looks like the build is still failing at least partially [due to format errors](https://circleci.com/gh/zio/zio/2001). Maybe you formatted the other one? I'd try again to format this one using `fmt` in `sbt`.

TheMayoras commented:
@jdegoes 

So it looks like 4 tests are failing?

I checked [jvm_jdk8](https://circleci.com/gh/zio/zio/2345?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) and it looks like it's failing a test where forked fibers are being supervised.  I checked to make sure all of the methods being used have the correct signatures, and they seems to be correct to me.

The other 3 tests...I have no idea.  When I open them up it says permission denied.

jdegoes commented:
@TheMayoras 

> I checked jvm_jdk8 and it looks like it's failing a test where forked fibers are being supervised. I checked to make sure all of the methods being used have the correct signatures, and they seems to be correct to me.

This is currently a flaky test. I can restart the build any time this happens.

Do you need any help this weekend to get this one through?

TheMayoras commented:
@jdegoes 

I'm not totally sure what all I have left other than figuring out why yarn isn't working.  Does anything jump out at you that I'm missing?

jdegoes commented:
> I'm not totally sure what all I have left other than figuring out why yarn isn't working. Does anything jump out at you that I'm missing?

It is hard to tell. There are old Circle CI failing tests.

Can you pull in upstream changes (`git pull upstream master`), and resolve?

Then can you do a `git rebase -i` to squash all commits into one? Or use a visual tool to do that? 

I think this will make the old "dead" Circle CI tasks go away, and then we can see if it's passing.

Let me know if you need any help!

TheMayoras commented:
@jdegoes 

I pulled in from upstream and rebased, but now I'm failing the lint test.  I ran `fmt` twice, so I'm not sure what else to try.

jdegoes commented:
There are still conflicts that must be resolved:

```
core/jvm/src/test/scala/zio/RTSSpec.scala
core/shared/src/main/scala/zio/ZIO.scala
```

Do you have time to work on it now?

TheMayoras commented:
@jdegoes

Yes, I have most of today to work on it

TheMayoras commented:
@jdegoes 

So I pulled in the changes, and I made sure to run `fmt`.  I'm still failing the linting.  Does the linter check for things other than the format?

TheMayoras commented:
@jdegoes 

I'm not quite sure what I'm doing to fail the linting, I've made sure to use `sbt fmt` and even reset my `JVMOTPS` and `SBTOPS` in case that was messing it up for some reason.  When I check the reason it fails, it says "permission denied"?

jdegoes commented:
@TheMayoras I should have some time tomorrow morning. Available?

TheMayoras commented:
@jdegoes 
Unfortunately, I am gone at work from about 8:20-6 (Eastern Time Zone).  I can try to meet before I leave, but if I recall you are on mountain time which would make that quite early.  I can always meet later or this weekend. If I am not at work, I will absolutely shift my schedule to meet!

jdegoes commented:
@TheMayoras Let's try to do Wednesday, if that works!

TheMayoras commented:
@jdegoes 
Absolutely, what time works for you?

jdegoes commented:
@TheMayoras I am available much of this afternoon, let me know if any time works for you!

TheMayoras commented:
@jdegoes 
I can do any time after 6 Eastern Time (I believe that is 4 your time).  Does 6:15 (4:15 MST) work?

jdegoes commented:
@TheMayoras Let's do 8pm your time (which is now my time, I'm in Maryland).

TheMayoras commented:
@jdegoes 
Absolutely, Zoom again?

jdegoes commented:
@TheMayoras Perfect!

jdegoes commented:
@TheMayoras https://zoom.us/j/679844581

TheMayoras commented:
@jdegoes 

The lint passed! 😁
Compile dotty failed 😕

jdegoes commented:
@TheMayoras 

A quick look shows this error:

```
[error] -- [E007] Type Mismatch Error: /home/circleci/project/core/shared/src/main/scala/zio/ZIO.scala:461:39 
[error] 461 |      ZIO.descriptorWith(descriptor => if (descriptor.interrupted) cleanup else ZIO.unit)
[error]     |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]     |Found:    zio.ZIO[R1, Nothing, _](cleanup) | zio.ZIO[Any, Nothing, Unit](zio.ZIO.unit)
[error]     |Required: zio.ZIO[R1, Nothing, Any]
[error]     |
[error]     |where:    R1 is a type in method onInterrupt with bounds <: R
```

The type of `cleanup` just needs to be changed to `cleanup: ZIO[R1, Nothing, Any]` instead of `_`. 

That should fix the Dotty error!

TheMayoras commented:
@jdegoes 

It looks like one of those fiber timeout errors caused the a Supervised test to fail.  I'm not sure how to re-run the test on CircleCI, but I believe it passed last time.

jdegoes commented:
@TheMayoras Restarted!

ghostdogpr commented:
@jdegoes in CircleCI instead of clicking “Rebuild” from the failed job (that restarts the whole thing), you can click on “build” below “Workflow”, then “Rerun > Rerun from failed”. It will restart only the failed jobs, which is much faster.

jdegoes commented:
@TheMayoras The last thing we should do is "sanity check" that we haven't lost anything (we lost a new implementation of `bracketExit`, the `_` => `Any` change).

I'm not sure the best way to do this.

 A quick search reveals [this tool](https://users.semanticmerge.com/Download) that has a [community contributed parser for Scala](https://github.com/sageserpent-open/SemanticMergeScalaPlugin). 

Failing an automated way, I can help eye-balling it if you like. 😄

jdegoes commented:
@ghostdogpr Thank you! 🙏

TheMayoras commented:
@jdegoes 

I'll look into it tonight as soon as I get home from work!  Any help would be greatly appreciated!

TheMayoras commented:
@jdegoes 

So I have gone through all of the companion object files for `IO`, `UIO`, `Task`, `TaskR`, and I now have a list of missing methods.  If I recall, some of the objects lost some methods (for example `UIO` lost fail).  

Some methods I'm assuming are from newer pull requests that I missed.  I'll put those in.  The text file with the missing files is attached.

[missing-files.txt](https://github.com/zio/zio/files/3384598/missing-files.txt)

jdegoes commented:
I went through and annotated with `-U` (unneeded) and `-N` (needed).


```
UIO:
    _1 -U
    _2 -U
    access -U
    accessM -U
    checkSupervised -N
    checkTraced -N
    environment -U
    fail -U
    fromFunction -U
    fromFunctionM -U
    firstSuccessOf -U
    haltWith -U
    identity -U
    provide -U
    require -U
    runtime -N
    sandbox -U
    sequence -N
    sequencePar -N
    sequenceParN -N
    superviseStatus -N
    swap -U
    traverse -N
    traverse_ -N
    traversePar -N
    traversePar_ -N
    traverseParN -N
    traverseParN_ -N
    unsandbox -U
    unsupervised -N
    when -N
    whenM -N

IO:
    _1 -U
    _2 -U
    access -U
    accessM -U
    checkSupervised -N
    checkTraced -N
    environment -U
    fail -N
    fromFunction -U
    fromFunctionM -U
    firstSuccessOf -N
    haltWith -N
    identity -U
    provide -U
    require -N
    runtime -N
    sandbox -N
    sequence -N
    sequencePar -N
    sequenceParN -N
    superviseStatus -N
    swap -U
    traverse -N
    traverse_ -N
    traversePar -N
    traversePar_ -N
    traverseParN -N
    traverseParN_ -N
    unsandbox -N
    unsupervised -N
    when -N
    whenM -N

Task:
    _1 -U
    _2 -U
    access -U
    accessM -U
    checkSupervised -N
    checkTraced -N
    environment -U
    fail -N
    fromFunctionM -U
    firstSuccessOf -N
    fromFunction -U
    fromFunctionM -U
    haltWith -N
    identity -U
    provide -U
    require -N
    runtime -N
    sequence -N
    sequencePar -N
    sequenceParN -N
    superviseStatus -N
    swap -U
    traverse -N
    traverse_ -N
    traversePar -N
    traversePar_ -N
    traverseParN -N
    traverseParN_ -N
    unsandbox -N
    unsupervised -N
    when -N
    whenM -N

TaskR:
    _1 -N
    _2 -N
    access -N
    accessM -N
    checkSupervised -N
    checkTraced -N
    environment -N
    fail -N
    firstSuccessOf -N
    fromFunction -N
    fromFunctionM -N
    haltWith -N
    identity -N
    provide -N
    require -N
    runtime -N
    sandbox -N
    sequence -N
    sequencePar -N
    sequenceParN -N
    superviseStatus -N
    swap -N
    traverse -N
    traverse_ -N
    traversePar -N
    traversePar_ -N
    traverseParN -N
    traverseParN_ -N
    unsandbox -N
    unsupervised -N
    when -N
    whenM -N
```

jdegoes commented:
@TheMayoras See above. ^^^

Last thing we need to do is make sure `ZIO.scala` has all the updates.

TheMayoras commented:
@jdegoes 

I'm hoping to have this done tonight!  I'm planning to just marathon through all of it.

TheMayoras commented:
@jdegoes 

Is `sandbox` a `ZIO` trait function?

TheMayoras commented:
@jdegoes 

Also, are `Task` and `TaskR` allowed to have unsandbox?  The method requires a `Cause[Throwable]`, but the type alias is for `Throwable`

jdegoes commented:
@TheMayoras 

> Is sandbox a ZIO trait function?

Yes, although I thought it was a `ZIO.*` (companion object) method too.

> Also, are Task and TaskR allowed to have unsandbox? The method requires a Cause[Throwable], but the type alias is for Throwable

Yes, they are allowed to have `unsandbox`, and they return `Cause[Throwable]` (for both `Task` and `TaskR`, `E = Throwable`).

TheMayoras commented:
@jdegoes

As far as I can tell, `sandbox` is currently only in the `ZIO` trait.

Also, I get type errors when adding unsandbox to the `Task` types. The input parameter requires `ZIO[..., Cause[Throwable], ...]`

I'm trying to fix a weird doc bug at the moment. It's giving me am ambiguous link for the `bracket` and `bracketExit` methods in the companion objects.  I'm looking into it, but I cannot get it working (as you can probably tell from my 100000 commits recently)

jdegoes commented:
> As far as I can tell, sandbox is currently only in the ZIO trait.

All right, we can ignore it, then, we only need to worry about the `ZIO.*` object methods.

> I'm trying to fix a weird doc bug at the moment. It's giving me am ambiguous link for the bracket and bracketExit methods in the companion objects. I'm looking into it, but I cannot get it working (as you can probably tell from my 100000 commits recently)

I will try to help tomorrow. Family day today!

TheMayoras commented:
@jdegoes 

So everything seems to be working well except for the docs.  

Besides that, I think everything is finished.

jdegoes commented:
@TheMayoras Awesome work! 🎉 

I had a chance to look at the `mdoc` errors:

```
[error] /home/circleci/project/core/shared/src/main/scala/zio/UIO.scala:30:3: The link target "zio.ZIO$.bracket" is ambiguous. Several members fit the target:
[error] [R, E, A, B](acquire: zio.ZIO[R,E,A],release: A => zio.ZIO[R, Nothing, _],use: A => zio.ZIO[R,E,B]): zio.ZIO[R,E,B] in object ZIO [chosen]
[error] [R, E, A](acquire: zio.ZIO[R,E,A]): zio.ZIO.BracketAcquire[R,E,A] in object ZIO
[error] 
[error] 
[error] Quick crash course on using Scaladoc links
[error] ==========================================
[error] Disambiguating terms and types: Prefix terms with '$' and types with '!' in case both names are in use:
[error]  - [[scala.collection.immutable.List!.apply class List's apply method]] and
[error]  - [[scala.collection.immutable.List$.apply object List's apply method]]
[error] Disambiguating overloaded members: If a term is overloaded, you can indicate the first part of its signature followed by *:
[error]  - [[[scala.collection.immutable.List$.fill[A](Int)(⇒A):List[A]* Fill with a single parameter]]]
[error]  - [[[scala.collection.immutable.List$.fill[A](Int,Int)(⇒A):List[List[A]]* Fill with a two parameters]]]
[error] Notes:
[error]  - you can use any number of matching square brackets to avoid interference with the signature
[error]  - you can use \\. to escape dots in prefixes (don't forget to use * at the end to match the signature!)
[error]  - you can use \\# to escape hashes, otherwise they will be considered as delimiters, like dots.
[error]   /**
[error]   ^
[error] /home/circleci/project/core/shared/src/main/scala/zio/UIO.scala:24:3: The link target "zio.ZIO$.bracket" is ambiguous. Several members fit the target:
[error] [R, E, A, B](acquire: zio.ZIO[R,E,A],release: A => zio.ZIO[R, Nothing, _],use: A => zio.ZIO[R,E,B]): zio.ZIO[R,E,B] in object ZIO [chosen]
[error] [R, E, A](acquire: zio.ZIO[R,E,A]): zio.ZIO.BracketAcquire[R,E,A] in object ZIO
[error] 
[error]   /**
[error]   ^
[error] /home/circleci/project/core/shared/src/main/scala/zio/UIO.scala:42:3: The link target "zio.ZIO$.bracketExit" is ambiguous. Several members fit the target:
[error] [R, E, A, B](acquire: zio.ZIO[R,E,A],release: (A, zio.Exit[E,B]) => zio.ZIO[R, Nothing, _],use: A => zio.ZIO[R,E,B]): zio.ZIO[R,E,B] in object ZIO [chosen]
[error] [R, E, A](acquire: zio.ZIO[R,E,A]): zio.ZIO.BracketExitAcquire[R,E,A] in object ZIO
[error] 
[error]   /**
[error]   ^
[error] /home/circleci/project/core/shared/src/main/scala/zio/UIO.scala:36:3: The link target "zio.ZIO$.bracketExit" is ambiguous. Several members fit the target:
[error] [R, E, A, B](acquire: zio.ZIO[R,E,A],release: (A, zio.Exit[E,B]) => zio.ZIO[R, Nothing, _],use: A => zio.ZIO[R,E,B]): zio.ZIO[R,E,B] in object ZIO [chosen]
[error] [R, E, A](acquire: zio.ZIO[R,E,A]): zio.ZIO.BracketExitAcquire[R,E,A] in object ZIO
[error] 
[error]   /**
[error]   ^
[error] /home/circleci/project/core/shared/src/main/scala/zio/TaskR.scala:46:3: Could not find any member to link for "zio.ZIO$.bracket[R,".
[error]   /**
[error]   ^
[error] /home/circleci/project/core/shared/src/main/scala/zio/TaskR.scala:40:3: Could not find any member to link for "zio.ZIO$.bracket[R,".
[error]   /**
[error]   ^
[error] /home/circleci/project/core/shared/src/main/scala/zio/TaskR.scala:61:3: Could not find any member to link for "zio.ZIO$.bracketExit[R,".
[error]   /**
[error]   ^
[error] /home/circleci/project/core/shared/src/main/scala/zio/TaskR.scala:55:3: Could not find any member to link for "zio.ZIO$.bracketExit[R,".
[error]   /**
[error]   ^
[error] /home/circleci/project/core/shared/src/main/scala/zio/Task.scala:32:3: Could not find any member to link for "zio.ZIO$.bracket[R,".
[error]   /**
[error]   ^
[error] /home/circleci/project/core/shared/src/main/scala/zio/Task.scala:26:3: Could not find any member to link for "zio.ZIO$.bracket[R,".
[error]   /**
[error]   ^
[error] /home/circleci/project/core/shared/src/main/scala/zio/Task.scala:44:3: Could not find any member to link for "zio.ZIO$.bracketExit[R,".
[error]   /**
[error]   ^
[error] /home/circleci/project/core/shared/src/main/scala/zio/Task.scala:38:3: Could not find any member to link for "zio.ZIO$.bracketExit[R,".
[error]   /**
[error]   ^
[error] /home/circleci/project/core/shared/src/main/scala/zio/IO.scala:32:3: Could not find any member to link for "zio.ZIO$.bracket[R,".
[error]   /**
[error]   ^
[error] /home/circleci/project/core/shared/src/main/scala/zio/IO.scala:26:3: Could not find any member to link for "zio.ZIO.bracket[R,".
[error]   /**
[error]   ^
[error] /home/circleci/project/core/shared/src/main/scala/zio/IO.scala:44:3: The link target "zio.ZIO$.bracketExit" is ambiguous. Several members fit the target:
[error] [R, E, A, B](acquire: zio.ZIO[R,E,A],release: (A, zio.Exit[E,B]) => zio.ZIO[R, Nothing, _],use: A => zio.ZIO[R,E,B]): zio.ZIO[R,E,B] in object ZIO [chosen]
[error] [R, E, A](acquire: zio.ZIO[R,E,A]): zio.ZIO.BracketExitAcquire[R,E,A] in object ZIO
[error] 
[error]   /**
[error]   ^
[error] /home/circleci/project/core/shared/src/main/scala/zio/IO.scala:38:3: The link target "zio.ZIO$.bracketExit" is ambiguous. Several members fit the target:
[error] [R, E, A, B](acquire: zio.ZIO[R,E,A],release: (A, zio.Exit[E,B]) => zio.ZIO[R, Nothing, _],use: A => zio.ZIO[R,E,B]): zio.ZIO[R,E,B] in object ZIO [chosen]
[error] [R, E, A](acquire: zio.ZIO[R,E,A]): zio.ZIO.BracketExitAcquire[R,E,A] in object ZIO
[error] 
[error]   /**
[error]   ^
[error] 16 errors found
[error] (coreJVM / Compile / doc) Scaladoc generation failed
[error] Total time: 39 s, completed Jul 14, 2019 10:16:45 PM
Exited with code 1
```

Here's what's happening: when linking to the ZIO object methods, via `[[methodReference]]` inside the Scaladoc of the `UIO` / `TaskR` / `Task` / `IO` companion objects, the documentation compiler finds multiple methods with the same name, and does not know which one to link to.

To disambiguate them, we can specify more details, like the above error message suggests.

Something like this should do the trick:

```
[[[zio.ZIO$.bracket[R, E, A](zio.ZIO[R,E,A]): zio.ZIO.BracketAcquire[R,E,A] ]]]
```

You might have to play around to get the syntax exactly right.

TheMayoras commented:
@jdegoes 

So I tried adding more details (like `See [[[[zio.ZIO.bracket[R, E, A, B]*]]]]`).  It always returns:

TheMayoras commented:
@jdegoes 

So I tried adding `See [[[[zio.ZIO.bracket[R, E, A, B]*]]]]`, but it always bring back:
`Could not find any member to link for "zio.ZIO$.bracket[R,"`.  
I feel like I'm not entering it correctly, but I found it from [here](https://stackoverflow.com/questions/23181672/cannot-create-scaladoc-link-for-overloaded-method).

TheMayoras commented:
Also, the tests seem to pass most of the time.  The failing jdk11 test passes most of the time.

jdegoes commented:
@TheMayoras How about for now, you just say:

```
see `ZIO.bracket`
```

and not use a Scaladoc link? 

Don't worry about the flaky tests, they are not caused by this PR. 😄

TheMayoras commented:
@jdegoes

I will fix it up when I get home at 6!  

Thank you so much for all of your help. I'm sure it's taken me longer than it should have, so thank you for your patience and help!

jdegoes commented:
@TheMayoras It's been a pleasure, and thank you for your help!

iravid commented on interop-reactiveStreams/shared/src/main/scala/zio/interop/reactiveStreams/QueueSubscriber.scala:
```diff
@@ -0,0 +1,91 @@
+package zio.interop.reactiveStreams
```
This file can be removed as it was moved to a separate repo

TheMayoras commented:
@jdegoes 

It looks like the only failing test is that flaky `Timeout` error.

jdegoes commented:
@TheMayoras Thanks for your immense work on this mega-project, and your patience sorting through all the details. And congratulations on your first ZIO pull request!

![giphy](https://user-images.githubusercontent.com/156745/61293409-9f8e7f00-a7a1-11e9-9722-469d1e99ca78.gif)

TheMayoras commented:
@jdegoes 

Thank you so much! It's been a journey, and I was more than a little scared to start with.  You are a saint for helping me so much!
